File: apps/web/src/api/lyrics.ts
/**
 * Lyrics from LRCLIB — free, no API key, around three million tracks.
 *
 * Fetched straight from the browser for now. When nyx-api lands it should
 * proxy and cache these, so a Pi with no internet still serves what it has
 * already seen. Until then, absence is a designed state, not an error.
 */

export interface LyricLine {
  /** Seconds from the start of the track. */
  time: number
  text: string
}

export type Lyrics =
  | { kind: 'synced'; lines: LyricLine[] }
  | { kind: 'plain'; text: string }
  | { kind: 'instrumental' }
  | { kind: 'absent' }

/**
 * Parse LRC.
 *
 * Real files are messier than the format suggests: a single line can carry
 * several timestamps, blank lines are meaningful pauses, and metadata tags
 * like [ar:] look exactly like timestamps until you check.
 */
export function parseLrc(lrc: string): LyricLine[] {
  const out: LyricLine[] = []
  const stamp = /\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]/g

  for (const raw of lrc.split(/\r?\n/)) {
    stamp.lastIndex = 0
    const times: number[] = []
    let m: RegExpExecArray | null
    while ((m = stamp.exec(raw)) !== null) {
      const min = Number(m[1])
      const sec = Number(m[2])
      // Fractions may be 1–3 digits: .5 is 500ms, .05 is 50ms, .050 is 50ms.
      const frac = m[3] ? Number(m[3]) / 10 ** m[3].length : 0
      times.push(min * 60 + sec + frac)
    }
    if (times.length === 0) continue

    const text = raw.replace(stamp, '').trim()
    // One line can be timestamped several times in a repeating chorus.
    for (const time of times) out.push({ time, text })
  }

  return out.sort((a, b) => a.time - b.time)
}

/** Index of the line active at `t`, or -1 before the first. */
export function activeLineAt(lines: LyricLine[], t: number): number {
  if (lines.length === 0) return -1
  // Binary search: this runs on every animation tick.
  let lo = 0
  let hi = lines.length - 1
  let ans = -1
  while (lo <= hi) {
    const mid = (lo + hi) >> 1
    if (lines[mid]!.time <= t) { ans = mid; lo = mid + 1 } else { hi = mid - 1 }
  }
  return ans
}

interface LrclibResponse {
  instrumental?: boolean
  plainLyrics?: string | null
  syncedLyrics?: string | null
}

export async function fetchLyrics(
  { title, artist, album, duration }:
  { title: string; artist: string; album: string; duration: number },
  signal?: AbortSignal,
): Promise<Lyrics> {
  const qs = new URLSearchParams({
    track_name: title,
    artist_name: artist,
    album_name: album,
    duration: String(Math.round(duration)),
  })

  const res = await fetch(`https://lrclib.net/api/get?${qs}`, signal ? { signal } : {})
  // 404 is the normal "nobody has transcribed this" answer, not a failure.
  if (res.status === 404) return { kind: 'absent' }
  if (!res.ok) throw new Error(`lrclib ${res.status}`)

  const data = await res.json() as LrclibResponse
  if (data.instrumental) return { kind: 'instrumental' }

  if (data.syncedLyrics) {
    const lines = parseLrc(data.syncedLyrics)
    if (lines.length > 0) return { kind: 'synced', lines }
  }
  if (data.plainLyrics?.trim()) return { kind: 'plain', text: data.plainLyrics }
  return { kind: 'absent' }
}

File: apps/api/src/nyx_api/lyrics.py
"""LRCLIB proxy and cache.

The client currently fetches lyrics straight from the browser, which means
nothing is remembered between sessions and an offline Pi serves nothing.
Caching here fixes both: once seen, a lyric is served from disk forever.
"""
from __future__ import annotations

import json
import re
import sqlite3
from datetime import datetime, timezone

import httpx

LRCLIB = "https://lrclib.net/api/get"
TIMEOUT = httpx.Timeout(6.0, connect=3.0)

_STAMP = re.compile(r"\[(\d{1,3}):(\d{2})(?:[.:](\d{1,3}))?\]")


def cache_key(title: str, artist: str, album: str, duration: int) -> str:
    # Duration is part of the key because lrclib matches on it: two recordings
    # of the same qawwali can differ by ten minutes and have different words.
    # The separator matters: joining with nothing lets ("ab", "c") collide
    # with ("a", "bc").
    return " :: ".join(
        [
            title.strip().lower(),
            artist.strip().lower(),
            album.strip().lower(),
            str(duration),
        ]
    )


def read_cache(conn: sqlite3.Connection, key: str) -> dict | None:
    row = conn.execute(
        "SELECT kind, payload FROM lyrics_cache WHERE key = ?", (key,)
    ).fetchone()
    if row is None:
        return None
    result: dict = {"kind": row["kind"], "cached": True}
    if row["payload"]:
        result.update(json.loads(row["payload"]))
    return result


def write_cache(conn: sqlite3.Connection, key: str, kind: str, payload: dict) -> None:
    conn.execute(
        "INSERT OR REPLACE INTO lyrics_cache (key, kind, payload, fetched_at) "
        "VALUES (?, ?, ?, ?)",
        (
            key,
            kind,
            json.dumps(payload) if payload else None,
            datetime.now(timezone.utc).isoformat(),
        ),
    )
    conn.commit()


def parse_lrc(lrc: str) -> list[dict]:
    """Parse LRC into timed lines.

    Mirrors the client's parser deliberately: a line may carry several
    timestamps, fraction digits are one to three and mean different things,
    and metadata tags look like timestamps until checked.
    """
    out: list[dict] = []

    for raw in lrc.splitlines():
        times = []
        for m in _STAMP.finditer(raw):
            frac = m.group(3)
            seconds = (
                int(m.group(1)) * 60
                + int(m.group(2))
                + (int(frac) / 10 ** len(frac) if frac else 0)
            )
            times.append(seconds)
        if not times:
            continue
        text = _STAMP.sub("", raw).strip()
        out.extend({"time": t, "text": text} for t in times)

    return sorted(out, key=lambda line: line["time"])


async def fetch(
    conn: sqlite3.Connection, title: str, artist: str, album: str, duration: int
) -> dict:
    key = cache_key(title, artist, album, duration)
    hit = read_cache(conn, key)
    if hit is not None:
        return hit

    params = {
        "track_name": title,
        "artist_name": artist,
        "album_name": album,
        "duration": duration,
    }
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            response = await client.get(LRCLIB, params=params)
    except httpx.HTTPError:
        # Offline. Do NOT cache this: absence of internet is not absence of
        # lyrics, and caching it would poison the entry permanently.
        return {"kind": "absent", "cached": False}

    if response.status_code == 404:
        # A real answer, worth caching: nobody has transcribed this.
        write_cache(conn, key, "absent", {})
        return {"kind": "absent", "cached": False}

    if response.status_code != 200:
        return {"kind": "absent", "cached": False}

    data = response.json()

    if data.get("instrumental"):
        write_cache(conn, key, "instrumental", {})
        return {"kind": "instrumental", "cached": False}

    synced = data.get("syncedLyrics")
    if synced:
        lines = parse_lrc(synced)
        if lines:
            write_cache(conn, key, "synced", {"lines": lines})
            return {"kind": "synced", "lines": lines, "cached": False}

    plain = data.get("plainLyrics")
    if plain and plain.strip():
        write_cache(conn, key, "plain", {"text": plain})
        return {"kind": "plain", "text": plain, "cached": False}

    write_cache(conn, key, "absent", {})
    return {"kind": "absent", "cached": False}
